core/src/main/java/org/kohsuke/stapler/export/DataWriter.java

```diff
@@ -53,6 +54,22 @@
     void startArray() throws IOException;
     void endArray() throws IOException;
 
+    /**
+     * Augments the next {@link #startObject()} call by specifying the type information of that object.
+     *
+     * @param expected
+     *      The declared type of the variable that references this object.
+     *      Null if the object is not referenced by anyone, for example when it's the root.
+     * @param actual
+     *      The actual type of the object being written.
+     *      Null if the object is synthetic is has no valid Java type
+     */
+    void type(Type expected, Class actual) throws IOException;
     void startObject() throws IOException;
```

Comment:
(deprecating `startObject()` but calling it as the fallback when we get `AbstractMethodError`; in `-source 8` could add a default implementation for `startObject(Type, Class)`)